src/module_wrap.cc

```diff
@@ -63,6 +62,66 @@ using v8::UnboundModuleScript;
 using v8::Undefined;
 using v8::Value;
 
+ModulePhase to_phase_constant(ModuleImportPhase phase) {
+  switch (phase) {
+    case ModuleImportPhase::kEvaluation:
+      return kEvaluationPhase;
+    case ModuleImportPhase::kSource:
+      return kSourcePhase;
+  }
+  UNREACHABLE();
+}
+
+void ModuleRequest::MemoryInfo(MemoryTracker* tracker) const {
+  tracker->TrackField("specifier", specifier);
+  tracker->TrackField("import_attributes", import_attributes);
+  tracker->TrackField("phase", static_cast<int>(phase));
+}
+
+template <int elements_per_attribute>
+ModuleRequest ModuleRequest::From(Local<Context> context,
+                                  Local<String> specifier,
+                                  Local<FixedArray> import_attributes,
+                                  ModulePhase phase) {
+  CHECK_EQ(import_attributes->Length() % elements_per_attribute, 0);
+  Isolate* isolate = context->GetIsolate();
+  std::size_t h1 = specifier->GetIdentityHash();
+  size_t num_attributes = import_attributes->Length() / elements_per_attribute;
+  ImportAttributeVector attributes;
+  attributes.reserve(num_attributes);
+
+  std::size_t h2 = 0;
+
+  for (int i = 0; i < import_attributes->Length();
+       i += elements_per_attribute) {
+    Local<String> v8_key = import_attributes->Get(context, i).As<String>();
+    Local<String> v8_value =
+        import_attributes->Get(context, i + 1).As<String>();
+    Utf8Value key_utf8(isolate, v8_key);
+    Utf8Value value_utf8(isolate, v8_value);
+
+    attributes.emplace_back(key_utf8.ToString(), value_utf8.ToString());
+    h2 ^= v8_key->GetIdentityHash();
+    h2 ^= v8_value->GetIdentityHash();
+  }
+
+  std::size_t h3 = std::hash<int>{}(static_cast<int>(phase));
```

Comment:
Formally, the key should be independent of the phase, so that `import(instance)` and `import(source)` can always be identified to be the same module. I hope we can work towards this concept of identity. We could possibly prepare the way by keeping the hash untouched by the phase...